pkg/cloud/gcp/actuators/machine/reconciler_test.go

```diff
@@ -1164,3 +1164,83 @@ func TestRestartPolicyToBool(t *testing.T) {
 		})
 	}
 }
+
+func TestEnsureCorrectNetworkAndSubnetName(t *testing.T) {
+	testType := "testType"
+	testRegion := "testRegion"
+	testZone := "testZone"
+
+	r := Reconciler{
+		machineScope: &machineScope{
+
+			machine: &machinev1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "",
+					Namespace: "",
+					Labels: map[string]string{
+						machinev1.MachineClusterIDLabel: "test-machine-1",
+						openshiftMachineRoleLabel:       "test-machine-role",
+					},
+				},
+			},
+			providerSpec: &machinev1.GCPMachineProviderSpec{
+				MachineType: testType,
+				Region:      testRegion,
+				Zone:        testZone,
+				NetworkInterfaces: []*machinev1.GCPNetworkInterface{
+					{
+						Network:    "custom-test-network",
+						Subnetwork: "custom-test-subnetwork",
+					},
+				},
+			},
+		},
+	}
+
+	cases := []struct {
+		name                   string
+		expectedNetworkName    string
+		expectedSubnetworkName string
+		expectedError          error
+	}{
+		{
+			name:                   "Custom network and subnetwork name",
+			expectedNetworkName:    "custom-test-network",
+			expectedSubnetworkName: "custom-test-subnetwork",
+			expectedError:          nil,
+		},
+		{
+			name:                   "Expected network and subnetwork name present",
+			expectedNetworkName:    "test-machine-1-network",
+			expectedSubnetworkName: "test-machine-1-test-machine-role-subnet",
+			expectedError:          nil,
+		},
+	}
+
+	for i, tc := range cases {
+		// Append the expected format of network and subnetwork according to GCP docs
+		if i == 1 {
+			r.providerSpec.NetworkInterfaces = append(r.providerSpec.NetworkInterfaces, &machinev1.GCPNetworkInterface{
+				Network:    "test-machine-1-network",
+				Subnetwork: "test-machine-1-test-machine-role-subnet",
+			})
+		}
```

Comment:
This is fragile and won't scale well, what happens if we want to include more cases later?

I would suggest instead to have a slice of network interfaces in your cases struct and set that within each test case, then the entire variation is included within the slice

I think we need several cases anyway:
* Single interface with unexpected values
* Single interface with expected values
* Two interfaces, expected in index 0
* Two interfaces, expected in index 1
Can you work on ensuring we have all of these cases?

Reply:
Good point. Will modify this.

Reply:
This won't be possible with the current implementation and I don't think that this is an issue.
This is also a good point for the current implementation IMO. If there is 20 networks and the one that we expect (the one that is always created by the installer) is on, let's say, 6th index, we don't want to just return 0th index of the `networkInterfaces`. I think that it would be better to just always put the machine in the basic network, unless the user decides otherwise by deleting this network resource - VPC use case.